File: backend/seed.py
import json
from datetime import date, timedelta
from decimal import Decimal
from pathlib import Path

from sqlalchemy import select

from app.core.database import SessionLocal
from app.core.security import get_password_hash
from app.core.storage import ensure_storage_dirs
from app.models.attendance import Attendance
from app.models.batch import Batch, StudentBatch
from app.models.enums import AttendanceStatus, FeePlanType, PaymentMode, StudentStatus, UserRole
from app.models.fee import FeePlan, Payment, ReminderRule, StudentFee
from app.models.institute import Institute
from app.models.note import Note
from app.models.student import Student
from app.models.user import User


def _create_sample_note_file(storage_dir: str, filename: str, content: str) -> str:
    notes_dir = Path(storage_dir) / "notes"
    notes_dir.mkdir(parents=True, exist_ok=True)
    file_path = notes_dir / filename
    file_path.write_text(content, encoding="utf-8")
    return str(Path("notes") / filename)


def run_seed() -> None:
    db = SessionLocal()
    ensure_storage_dirs()
    try:
        existing_user = db.scalar(select(User.id).limit(1))
        if existing_user:
            print("Seed skipped: data already exists.")
            return

        institute = Institute(name="Demo Coaching Institute")
        db.add(institute)
        db.flush()

        admin = User(
            institute_id=institute.id,
            full_name="System Admin",
            email="admin@demo.com",
            phone="9990000001",
            password_hash=get_password_hash("Admin@123"),
            role=UserRole.ADMIN,
            is_active=True,
        )
        teacher = User(
            institute_id=institute.id,
            full_name="Anita Sharma",
            email="teacher@demo.com",
            phone="9990000002",
            password_hash=get_password_hash("Teacher@123"),
            role=UserRole.TEACHER,
            is_active=True,
        )
        db.add_all([admin, teacher])
        db.flush()

        students: list[Student] = []
        for index in range(1, 6):
            student = Student(
                institute_id=institute.id,
                full_name=f"Student {index}",
                phone=f"900000000{index}",
                email=f"student{index}@demo.com",
                guardian_name=f"Guardian {index}",
                guardian_phone=f"800000000{index}",
                address=f"City Block {index}",
                join_date=date.today() - timedelta(days=60 - index),
                status=StudentStatus.ACTIVE,
            )
            db.add(student)
            db.flush()
            students.append(student)
            db.add(
                User(
                    institute_id=institute.id,
                    full_name=student.full_name,
                    email=student.email,
                    phone=student.phone,
                    password_hash=get_password_hash("Student@123"),
                    role=UserRole.STUDENT,
                    is_active=True,
                    student_id=student.id,
                )
            )

        monthly_plan = FeePlan(
            institute_id=institute.id,
            name="Monthly Standard",
            type=FeePlanType.MONTHLY,
            amount=Decimal("3000.00"),
            metadata_json=json.dumps({"months": 3}),
        )
        quarterly_plan = FeePlan(
            institute_id=institute.id,
            name="Quarterly Pro",
            type=FeePlanType.QUARTERLY,
            amount=Decimal("8500.00"),
            metadata_json=json.dumps({"includes_material": True}),
        )
        db.add_all([monthly_plan, quarterly_plan])
        db.flush()

        batch_1 = Batch(
            institute_id=institute.id,
            name="Class 10 Maths Morning",
            course="Class 10 Mathematics",
            schedule="Mon-Wed-Fri 07:00 AM - 08:30 AM",
            teacher_id=teacher.id,
            start_date=date.today() - timedelta(days=30),
            end_date=date.today() + timedelta(days=120),
            fee_plan_id=monthly_plan.id,
        )
        batch_2 = Batch(
            institute_id=institute.id,
            name="Class 12 Physics Evening",
            course="Class 12 Physics",
            schedule="Tue-Thu-Sat 06:00 PM - 07:30 PM",
            teacher_id=teacher.id,
            start_date=date.today() - timedelta(days=20),
            end_date=date.today() + timedelta(days=140),
            fee_plan_id=quarterly_plan.id,
        )
        db.add_all([batch_1, batch_2])
        db.flush()

        for student in students[:3]:
            db.add(StudentBatch(institute_id=institute.id, student_id=student.id, batch_id=batch_1.id))
        for student in students[2:]:
            db.add(StudentBatch(institute_id=institute.id, student_id=student.id, batch_id=batch_2.id))

        note_path_1 = _create_sample_note_file(
            storage_dir=str(Path(__file__).resolve().parent / "storage"),
            filename="algebra-intro.txt",
            content="Algebra fundamentals and sample solved examples.",
        )
        note_path_2 = _create_sample_note_file(
            storage_dir=str(Path(__file__).resolve().parent / "storage"),
            filename="physics-motion.txt",
            content="Kinematics notes and practice worksheet references.",
        )
        db.add_all(
            [
                Note(
                    institute_id=institute.id,
                    batch_id=batch_1.id,
                    title="Algebra Revision Sheet",
                    description="Chapter-wise revision notes.",
                    tags="algebra,revision",
                    file_name="algebra-intro.txt",
                    file_path=note_path_1,
                    file_type="text/plain",
                    created_by=teacher.id,
                ),
                Note(
                    institute_id=institute.id,
                    batch_id=batch_2.id,
                    title="Physics Motion Notes",
                    description="Intro to motion and vectors.",
                    tags="physics,motion",
                    file_name="physics-motion.txt",
                    file_path=note_path_2,
                    file_type="text/plain",
                    created_by=teacher.id,
                ),
            ]
        )

        for day_delta in range(7):
            attendance_date = date.today() - timedelta(days=day_delta + 1)
            for student in students[:3]:
                db.add(
                    Attendance(
                        institute_id=institute.id,
                        batch_id=batch_1.id,
                        student_id=student.id,
                        date=attendance_date,
                        status=AttendanceStatus.PRESENT if (student.id + day_delta) % 4 != 0 else AttendanceStatus.ABSENT,
                        marked_by=teacher.id,
                    )
                )
            for student in students[2:]:
                db.add(
                    Attendance(
                        institute_id=institute.id,
                        batch_id=batch_2.id,
                        student_id=student.id,
                        date=attendance_date,
                        status=AttendanceStatus.PRESENT if (student.id + day_delta) % 5 != 0 else AttendanceStatus.ABSENT,
                        marked_by=teacher.id,
                    )
                )

        def schedule_for(base_amount: Decimal) -> str:
            schedule = [
                {"due_date": (date.today() - timedelta(days=10)).isoformat(), "amount": str(base_amount / 3)},
                {"due_date": (date.today() + timedelta(days=5)).isoformat(), "amount": str(base_amount / 3)},
                {"due_date": (date.today() + timedelta(days=35)).isoformat(), "amount": str(base_amount / 3)},
            ]
            return json.dumps(schedule)

        student_fee_rows: list[StudentFee] = []
        for student in students[:3]:
            row = StudentFee(
                institute_id=institute.id,
                student_id=student.id,
                batch_id=batch_1.id,
                fee_plan_id=monthly_plan.id,
                total_fee=Decimal("9000.00"),
                discount=Decimal("500.00"),
                due_schedule_json=schedule_for(Decimal("9000.00")),
            )
            db.add(row)
            student_fee_rows.append(row)

        for student in students[2:]:
            row = StudentFee(
                institute_id=institute.id,
                student_id=student.id,
                batch_id=batch_2.id,
                fee_plan_id=quarterly_plan.id,
                total_fee=Decimal("12000.00"),
                discount=Decimal("1000.00"),
                due_schedule_json=schedule_for(Decimal("12000.00")),
            )
            db.add(row)
            student_fee_rows.append(row)

        db.flush()

        for index, student_fee in enumerate(student_fee_rows[:4], start=1):
            db.add(
                Payment(
                    institute_id=institute.id,
                    student_fee_id=student_fee.id,
                    amount=Decimal("2500.00"),
                    paid_on=(date.today() - timedelta(days=2)).isoformat(),
                    mode=PaymentMode.UPI,
                    receipt_no=f"SEED-RCPT-{index:03d}",
                    remarks="Seed payment",
                    created_by=admin.id,
                )
            )

        db.add(
            ReminderRule(
                institute_id=institute.id,
                batch_id=None,
                name="Default Reminder Rule",
                days_before=3,
                on_due_date=True,
                every_n_days_after_due=2,
                is_active=True,
            )
        )

        db.commit()
        print("Seed completed.")
        print("Admin login: admin@demo.com / Admin@123")
        print("Teacher login: teacher@demo.com / Teacher@123")
        print("Student login: student1@demo.com / Student@123")
    except Exception:
        db.rollback()
        raise
    finally:
        db.close()


if __name__ == "__main__":
    run_seed()

File: backend/src/seed.js
const fs = require("fs");
const path = require("path");
const { initializeSchema, connectDb } = require("./db");
const { hashPassword } = require("./auth");
const { ensureStorageDirs } = require("./services");

function isoDateOffset(days) {
  const d = new Date();
  d.setDate(d.getDate() + days);
  return d.toISOString().slice(0, 10);
}

async function runSeed() {
  await initializeSchema();
  ensureStorageDirs();
  const db = await connectDb();

  const existing = await db.get("SELECT id FROM users LIMIT 1");
  if (existing) {
    process.stdout.write("Seed skipped: data already exists.\n");
    return;
  }

  const institute = await db.run("INSERT INTO institutes (name) VALUES (?)", ["Demo Coaching Institute"]);
  const instituteId = institute.lastID;

  const admin = await db.run(
    `INSERT INTO users (institute_id, full_name, email, phone, password_hash, role, is_active)
     VALUES (?, ?, ?, ?, ?, 'ADMIN', 1)`,
    [instituteId, "System Admin", "admin@demo.com", "9990000001", hashPassword("Admin@123")]
  );
  const teacher = await db.run(
    `INSERT INTO users (institute_id, full_name, email, phone, password_hash, role, is_active)
     VALUES (?, ?, ?, ?, ?, 'TEACHER', 1)`,
    [instituteId, "Anita Sharma", "teacher@demo.com", "9990000002", hashPassword("Teacher@123")]
  );

  const studentIds = [];
  for (let i = 1; i <= 5; i += 1) {
    const student = await db.run(
      `INSERT INTO students (institute_id, full_name, phone, email, guardian_name, guardian_phone, address, join_date, status)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'ACTIVE')`,
      [
        instituteId,
        `Student ${i}`,
        `900000000${i}`,
        `student${i}@demo.com`,
        `Guardian ${i}`,
        `800000000${i}`,
        `City Block ${i}`,
        isoDateOffset(-60 + i)
      ]
    );
    studentIds.push(student.lastID);
    await db.run(
      `INSERT INTO users (institute_id, full_name, email, phone, password_hash, role, student_id, is_active)
       VALUES (?, ?, ?, ?, ?, 'STUDENT', ?, 1)`,
      [
        instituteId,
        `Student ${i}`,
        `student${i}@demo.com`,
        `900000000${i}`,
        hashPassword("Student@123"),
        student.lastID
      ]
    );
  }

  const monthlyPlan = await db.run(
    `INSERT INTO fee_plans (institute_id, name, type, amount, metadata_json)
     VALUES (?, 'Monthly Standard', 'MONTHLY', 3000, ?)`,
    [instituteId, JSON.stringify({ months: 3 })]
  );
  const quarterlyPlan = await db.run(
    `INSERT INTO fee_plans (institute_id, name, type, amount, metadata_json)
     VALUES (?, 'Quarterly Pro', 'QUARTERLY', 8500, ?)`,
    [instituteId, JSON.stringify({ includes_material: true })]
  );

  const batch1 = await db.run(
    `INSERT INTO batches (institute_id, name, course, schedule, teacher_id, start_date, end_date, fee_plan_id)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      instituteId,
      "Class 10 Maths Morning",
      "Class 10 Mathematics",
      "Mon-Wed-Fri 07:00 AM - 08:30 AM",
      teacher.lastID,
      isoDateOffset(-30),
      isoDateOffset(120),
      monthlyPlan.lastID
    ]
  );
  const batch2 = await db.run(
    `INSERT INTO batches (institute_id, name, course, schedule, teacher_id, start_date, end_date, fee_plan_id)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      instituteId,
      "Class 12 Physics Evening",
      "Class 12 Physics",
      "Tue-Thu-Sat 06:00 PM - 07:30 PM",
      teacher.lastID,
      isoDateOffset(-20),
      isoDateOffset(140),
      quarterlyPlan.lastID
    ]
  );

  for (let i = 0; i < 3; i += 1) {
    await db.run("INSERT INTO student_batches (institute_id, student_id, batch_id) VALUES (?, ?, ?)", [
      instituteId,
      studentIds[i],
      batch1.lastID
    ]);
  }
  for (let i = 2; i < 5; i += 1) {
    await db.run("INSERT INTO student_batches (institute_id, student_id, batch_id) VALUES (?, ?, ?)", [
      instituteId,
      studentIds[i],
      batch2.lastID
    ]);
  }

  const notesDir = path.join(path.resolve(__dirname, ".."), "storage", "notes");
  fs.mkdirSync(notesDir, { recursive: true });
  fs.writeFileSync(path.join(notesDir, "algebra-intro.txt"), "Algebra fundamentals and solved examples.");
  fs.writeFileSync(path.join(notesDir, "physics-motion.txt"), "Kinematics notes and worksheet references.");

  await db.run(
    `INSERT INTO notes (institute_id, batch_id, title, description, tags, file_name, file_path, file_type, created_by)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      instituteId,
      batch1.lastID,
      "Algebra Revision Sheet",
      "Chapter-wise revision notes.",
      "algebra,revision",
      "algebra-intro.txt",
      "notes/algebra-intro.txt",
      "text/plain",
      teacher.lastID
    ]
  );
  await db.run(
    `INSERT INTO notes (institute_id, batch_id, title, description, tags, file_name, file_path, file_type, created_by)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      instituteId,
      batch2.lastID,
      "Physics Motion Notes",
      "Intro to motion and vectors.",
      "physics,motion",
      "physics-motion.txt",
      "notes/physics-motion.txt",
      "text/plain",
      teacher.lastID
    ]
  );

  for (let dayDelta = 1; dayDelta <= 7; dayDelta += 1) {
    const attendanceDate = isoDateOffset(-dayDelta);
    for (let i = 0; i < 3; i += 1) {
      await db.run(
        `INSERT INTO attendance (institute_id, batch_id, student_id, date, status, marked_by)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [
          instituteId,
          batch1.lastID,
          studentIds[i],
          attendanceDate,
          (studentIds[i] + dayDelta) % 4 !== 0 ? "PRESENT" : "ABSENT",
          teacher.lastID
        ]
      );
    }
    for (let i = 2; i < 5; i += 1) {
      await db.run(
        `INSERT INTO attendance (institute_id, batch_id, student_id, date, status, marked_by)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [
          instituteId,
          batch2.lastID,
          studentIds[i],
          attendanceDate,
          (studentIds[i] + dayDelta) % 5 !== 0 ? "PRESENT" : "ABSENT",
          teacher.lastID
        ]
      );
    }
  }

  function dueSchedule(total) {
    const installment = (total / 3).toFixed(2);
    return JSON.stringify([
      { due_date: isoDateOffset(-10), amount: installment },
      { due_date: isoDateOffset(5), amount: installment },
      { due_date: isoDateOffset(35), amount: installment }
    ]);
  }

  const studentFeeIds = [];
  for (let i = 0; i < 3; i += 1) {
    const row = await db.run(
      `INSERT INTO student_fees (institute_id, student_id, batch_id, fee_plan_id, total_fee, discount, due_schedule_json)
       VALUES (?, ?, ?, ?, 9000, 500, ?)`,
      [instituteId, studentIds[i], batch1.lastID, monthlyPlan.lastID, dueSchedule(9000)]
    );
    studentFeeIds.push(row.lastID);
  }
  for (let i = 2; i < 5; i += 1) {
    const row = await db.run(
      `INSERT INTO student_fees (institute_id, student_id, batch_id, fee_plan_id, total_fee, discount, due_schedule_json)
       VALUES (?, ?, ?, ?, 12000, 1000, ?)`,
      [instituteId, studentIds[i], batch2.lastID, quarterlyPlan.lastID, dueSchedule(12000)]
    );
    studentFeeIds.push(row.lastID);
  }

  for (let i = 0; i < 4; i += 1) {
    await db.run(
      `INSERT INTO payments (institute_id, student_fee_id, amount, paid_on, mode, receipt_no, remarks, created_by)
       VALUES (?, ?, 2500, ?, 'UPI', ?, 'Seed payment', ?)`,
      [instituteId, studentFeeIds[i], isoDateOffset(-2), `SEED-RCPT-${String(i + 1).padStart(3, "0")}`, admin.lastID]
    );
  }

  await db.run(
    `INSERT INTO reminder_rules (institute_id, batch_id, name, days_before, on_due_date, every_n_days_after_due, is_active)
     VALUES (?, NULL, 'Default Reminder Rule', 3, 1, 2, 1)`,
    [instituteId]
  );

  process.stdout.write("Seed completed.\n");
  process.stdout.write("Admin login: admin@demo.com / Admin@123\n");
  process.stdout.write("Teacher login: teacher@demo.com / Teacher@123\n");
  process.stdout.write("Student login: student1@demo.com / Student@123\n");
}

runSeed().catch((error) => {
  process.stderr.write(`Seed failed: ${error.message}\n`);
  process.exit(1);
});
